Fix a deadlock

weiqiangt opened the pull request:
When OFPBFC_TIMEOUT happened, listenReply would return and
there was no other consumer for `tx.controlReplyCh`, and OVS
would also send a timeout message to agent so publishMessage
would be called. However, the channel hadn't been removed
from txChans so it would be stuck in `ch <- result` forever.

This commit ensures controlReplyCh is always available to
be pushed to fix deadlock. controlReplyCh and intCh are also
added a buffer to reduce timeout wait impact.

Signed-off-by: Weiqiang Tang <weiqiangt@vmware.com>

weiqiangt committed:
Fix a deadlock race condition

There may be a deadlock ofswitch.txLock due to timer
difference. This commit makes receiver wait a bit longer
than OVS side to handle this issue. Also, when message
publishing costs 2x waitTimeout we just drop the message.

Signed-off-by: Weiqiang Tang <weiqiangt@vmware.com>

tnqn commented on ofctrl/ofSwitch.go:
```diff
@@ -460,7 +462,15 @@ func (self *OFSwitch) publishMessage(xID uint32, result MessageResult) {
-			ch <- result
+			select {
+			case ch <- result:
+			case <-time.After(messageTimeout):
```
Thanks for the quick fix. I still have concern that the routine blocks for 13 seconds with a lock, which could lead to other functions cannot proceed, it's like a JVM stop-the-world and unpredictable. Could we confirm whether a buffered channel can avoid blocking here? For example, if we can confirm the outbound and inbound message is always 1 to 1 interaction, then 1 buffer is enough and it doesn't need to wait?

tnqn replied:
There are several things to consider:
1. the channel should not be closed when it's in the map
2. the channel should have active consumer when `publishMessage` is adding items to it
3. `listenReply` (the consumer of the channel) should terminate eventually regardless the bundle is aborted or completed

So I'm thinking:
1. In `Transaction.Begin/Abort/Commit`, it should always unsubscribe first, then close the `controlReplyCh` channel, so that `publishMessage` won't send on closed channel
2. In `Transaction.listenReply`, it should always wait for the `controlReplyCh` channel to be closed first, so that the channel always has active consumer
3. `publishMessage` could be kept as is because as long as the channel is present, it's guaranteed not closed and having consumer.
4. `controlReplyCh` could have some buffer to avoid `publishMessage` holds the lock too long.
5. `controlIntCh` don't need to be closed as no consumer is waiting for that, and `listenReply` might send message to it.

weiqiangt replied:
Thanks for the suggestions. I've updated the code.

weiqiangt committed:
Fix a deadlock

When OFPBFC_TIMEOUT happened, listenReply would return and
there was no other consumer for `tx.controlReplyCh`, and OVS
would also send a timeout message to agent so publishMessage
would be called. However, the channel hadn't been removed
from txChans so it would be stuck in `ch <- result` forever.

This commit ensures controlReplyCh is always available to
be pushed to fix deadlock. controlReplyCh and intCh are also
added a buffer to reduce timeout wait impact.
This commit also polished go.mod.

Signed-off-by: Weiqiang Tang <weiqiangt@vmware.com>

tnqn commented on go.mod:
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/cenkalti/rpc2 v0.0.0-20140912135055-44d0d95e4f52 // indirect
 	github.com/contiv/libOpenflow v0.0.0-20200728044739-7c6534390721
 	github.com/contiv/libovsdb v0.0.0-20170227191248-d0061a53e358
+	github.com/google/martian v2.1.0+incompatible
```
I think the project uses "github.com/sirupsen/logrus"

weiqiangt replied:
Thanks, I didn't realize that the auto-import used this one.

tnqn commented on ofctrl/transaction.go:
```diff
@@ -6,13 +6,13 @@ package ofctrl
 
 	"github.com/contiv/libOpenflow/openflow13"
 	"github.com/contiv/libOpenflow/util"
+	"github.com/google/martian/log"
```
all other logs are using `log "github.com/sirupsen/logrus"`

tnqn reviewed (commented):
LGTM, except the inconsistent log import

tnqn commented on go.mod:
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/cenkalti/hub v1.0.1-0.20140529221144-7be60e186e66 // indirect
 	github.com/cenkalti/rpc2 v0.0.0-20140912135055-44d0d95e4f52 // indirect
-	github.com/contiv/libOpenflow v0.0.0-20200728044739-7c6534390721
+	github.com/contiv/libOpenflow v0.0.0-20201014051314-c1702744526c
```
Is this upgrade by intention? what change it brings?

weiqiangt replied:
There are breaking changes libOpenflow which will make the compile failed if we don't update here.

tnqn commented on ofctrl/ofSwitch.go:
```diff
@@ -32,7 +32,9 @@ import (
-	messageTimeout = 10 * time.Second
+	// The default BUNDLE_IDLE_TIMEOUT is 10s and we need wait a bit
+	// longer to overcome the timer difference.
+	messageTimeout = 13 * time.Second
```
Does it still need to be updated?

weiqiangt replied:
Since the timers will be slightly different I prefer to wait longer here.
But no strong opinion here.

tnqn reviewed (approved):
LGTM
